webapp/channels/src/packages/mattermost-redux/src/reducers/entities/posts.ts

```diff
@@ -200,11 +202,43 @@ export function handlePosts(state: RelationOneToOne<Post, Post> = {}, action: Ge
             },
         };
 
-        // Remove any of its comments
         for (const otherPost of Object.values(state)) {
+            // Remove any of its comments
             if (otherPost.root_id === post.id) {
                 Reflect.deleteProperty(nextState, otherPost.id);
             }
+
+            // a deleted post may exist in some other post's
+            // embeds when its link is mentioned in the post message.
+            // We need to remove the deleted post from post embeds of all posts
+            // to ensure the deleted post's contents cannot be retrieved from the store.
+            if (otherPost.metadata && otherPost.metadata.embeds && otherPost.metadata.embeds.length > 0) {
+                // This will become the post's new embeds array.
+                // We'll add everything other than the deleted post's embed here.
+                const newEmbeds: PostEmbed[] = [];
+
+                for (const embed of otherPost.metadata.embeds) {
+                    if (embed.type === 'permalink' && (embed.data as PostPreviewMetadata).post_id === post.id) {
+                        // skip if the embed is the deleted post
+                    } else {
+                        // include everything else
+                        newEmbeds.push(embed);
+                    }
```

Comment:
Personal preference, so feel free to keep it as you find it more readable.

```suggestion
                    if (embed.type === 'permalink' && (embed.data as PostPreviewMetadata).post_id === post.id) {
                        // skip if the embed is the deleted post
                        continue;
                    }
                    
                    // include everything else
                    newEmbeds.push(embed);
```